File: request/lib/src/services/sms_auth_service.dart
import 'package:http/http.dart' as http;
import 'dart:convert';
import 'package:cloud_functions/cloud_functions.dart';

class SMSAuthService {
  static final SMSAuthService _instance = SMSAuthService._internal();
  factory SMSAuthService() => _instance;
  SMSAuthService._internal();

  final FirebaseFunctions _functions = FirebaseFunctions.instanceFor(region: 'us-central1');

  /// Check if user exists by email or phone
  Future<Map<String, dynamic>> checkUserExists(String emailOrPhone) async {
    try {
      final callable = _functions.httpsCallable('checkUserExists');
      final result = await callable.call({
        'emailOrPhone': emailOrPhone,
      });
      
      return {
        'exists': result.data['exists'] ?? false,
        'isEmail': result.data['inputType'] == 'email',
        'isPhone': result.data['inputType'] == 'phone',
        'userId': result.data['userId'],
      };
    } catch (e) {
      throw Exception('Failed to check user existence: $e');
    }
  }

  /// Send OTP for new user registration
  Future<Map<String, dynamic>> sendRegistrationOTP(String emailOrPhone) async {
    try {
      final callable = _functions.httpsCallable('sendRegistrationOTP');
      final result = await callable.call({
        'emailOrPhone': emailOrPhone,
      });
      
      return {
        'success': result.data['success'] ?? false,
        'message': result.data['message'] ?? '',
        'otpId': result.data['otpId'],
        'isEmail': result.data['isEmail'] ?? false,
        'expiresIn': result.data['expiresIn'] ?? 300,
      };
    } catch (e) {
      throw Exception('Failed to send registration OTP: $e');
    }
  }

  /// Send OTP for password reset
  Future<Map<String, dynamic>> sendPasswordResetOTP(String emailOrPhone) async {
    try {
      final callable = _functions.httpsCallable('sendPasswordResetOTP');
      final result = await callable.call({
        'emailOrPhone': emailOrPhone,
      });
      
      return {
        'success': result.data['success'] ?? false,
        'message': result.data['message'] ?? '',
        'otpId': result.data['otpId'],
        'isEmail': result.data['isEmail'] ?? false,
        'expiresIn': result.data['expiresIn'] ?? 300,
      };
    } catch (e) {
      throw Exception('Failed to send password reset OTP: $e');
    }
  }

  /// Verify OTP code
  Future<Map<String, dynamic>> verifyOTP({
    required String emailOrPhone,
    required String otp,
    required String otpId,
    required String purpose, // 'registration' or 'password_reset'
  }) async {
    try {
      final callable = _functions.httpsCallable('verifyOTP');
      final result = await callable.call({
        'emailOrPhone': emailOrPhone,
        'otp': otp,
        'otpId': otpId,
        'purpose': purpose,
      });
      
      return {
        'success': result.data['success'] ?? false,
        'message': result.data['message'] ?? '',
        'customToken': result.data['customToken'],
        'user': result.data['user'],
        'isNewUser': result.data['isNewUser'] ?? false,
      };
    } catch (e) {
      throw Exception('Failed to verify OTP: $e');
    }
  }

  /// Login with email/phone and password
  Future<Map<String, dynamic>> loginWithPassword({
    required String emailOrPhone,
    required String password,
  }) async {
    try {
      final callable = _functions.httpsCallable('loginWithPassword');
      final result = await callable.call({
        'emailOrPhone': emailOrPhone,
        'password': password,
      });
      
      return {
        'success': result.data['success'] ?? false,
        'message': result.data['message'] ?? '',
        'customToken': result.data['customToken'],
        'user': result.data['user'],
      };
    } catch (e) {
      throw Exception('Failed to login: $e');
    }
  }

  /// Complete profile for new user
  Future<Map<String, dynamic>> completeProfile({
    required String firstName,
    required String lastName,
    required String password,
    required String emailOrPhone,
    required String otpId,
  }) async {
    try {
      final callable = _functions.httpsCallable('completeProfile');
      final result = await callable.call({
        'firstName': firstName,
        'lastName': lastName,
        'password': password,
        'emailOrPhone': emailOrPhone,
        'otpId': otpId,
      });
      
      return {
        'success': result.data['success'] ?? false,
        'message': result.data['message'] ?? '',
        'customToken': result.data['customToken'],
        'user': result.data['user'],
      };
    } catch (e) {
      throw Exception('Failed to complete profile: $e');
    }
  }

  /// Reset password after OTP verification
  Future<Map<String, dynamic>> resetPassword({
    required String emailOrPhone,
    required String newPassword,
    required String otpId,
  }) async {
    try {
      final callable = _functions.httpsCallable('resetPassword');
      final result = await callable.call({
        'emailOrPhone': emailOrPhone,
        'newPassword': newPassword,
        'otpId': otpId,
      });
      
      return {
        'success': result.data['success'] ?? false,
        'message': result.data['message'] ?? '',
      };
    } catch (e) {
      throw Exception('Failed to reset password: $e');
    }
  }

  /// Detect country from phone number
  String detectCountryFromPhone(String phoneNumber) {
    if (phoneNumber.startsWith('+94')) return 'LK'; // Sri Lanka
    if (phoneNumber.startsWith('+91')) return 'IN'; // India
    if (phoneNumber.startsWith('+1')) return 'US';   // USA
    if (phoneNumber.startsWith('+44')) return 'UK';  // UK
    // Add more countries as needed
    return 'LK'; // Default to Sri Lanka
  }

  /// Validate email format
  bool isValidEmail(String email) {
    return RegExp(r'^[\w-\.]+@([\w-]+\.)+[\w-]{2,4}$').hasMatch(email);
  }

  /// Validate phone number format
  bool isValidPhone(String phone) {
    // Remove spaces and hyphens
    String cleanPhone = phone.replaceAll(RegExp(r'[\s-]'), '');
    // Check if it starts with + and has 10-15 digits
    return RegExp(r'^\+[1-9]\d{9,14}$').hasMatch(cleanPhone);
  }

  /// Format phone number
  String formatPhoneNumber(String phone) {
    // Remove all non-digit characters except +
    String cleaned = phone.replaceAll(RegExp(r'[^\d+]'), '');
    
    // Ensure it starts with +
    if (!cleaned.startsWith('+')) {
      // Assume Sri Lankan number if no country code
      if (cleaned.startsWith('0')) {
        cleaned = '+94${cleaned.substring(1)}';
      } else {
        cleaned = '+94$cleaned';
      }
    }
    
    return cleaned;
  }
}

File: admin-react/src/services/smsAuthService.js
/**
 * Custom SMS Authentication Service
 * 
 * @description
 * Client-side authentication service that replaces Firebase Auth with
 * custom SMS OTP authentication. Provides country-wise SMS providers
 * for cost optimization while maintaining security.
 * 
 * @features
 * - Custom SMS OTP authentication
 * - Country-specific SMS provider support
 * - Secure token management
 * - Rate limiting and retry logic
 * - Cost-effective authentication flow
 * 
 * @cost_benefits
 * - Reduces authentication costs by 50-80%
 * - No Firebase Auth monthly base fees
 * - Use of local/regional SMS providers
 * 
 * @author Request Marketplace Team
 * @version 1.0.0
 * @since 2025-08-16
 */

import { initializeApp } from 'firebase/app';
import { 
  getAuth, 
  signInWithCustomToken, 
  signOut as firebaseSignOut,
  onAuthStateChanged 
} from 'firebase/auth';
import { 
  getFirestore, 
  doc, 
  getDoc, 
  setDoc, 
  updateDoc,
  collection,
  query,
  where,
  getDocs,
  Timestamp 
} from 'firebase/firestore';
import { getFunctions, httpsCallable } from 'firebase/functions';

// Firebase configuration
const firebaseConfig = {
  // Your Firebase config here
  projectId: 'request-marketplace'
};

// Initialize Firebase
const app = initializeApp(firebaseConfig);
const auth = getAuth(app);
const db = getFirestore(app);
const functions = getFunctions(app);

// === AUTHENTICATION STATE MANAGEMENT ===

class AuthState {
  constructor() {
    this.user = null;
    this.isLoading = true;
    this.listeners = [];
  }

  setUser(user) {
    this.user = user;
    this.isLoading = false;
    this.notifyListeners();
  }

  setLoading(loading) {
    this.isLoading = loading;
    this.notifyListeners();
  }

  addListener(callback) {
    this.listeners.push(callback);
    return () => {
      this.listeners = this.listeners.filter(listener => listener !== callback);
    };
  }

  notifyListeners() {
    this.listeners.forEach(listener => {
      listener({
        user: this.user,
        isLoading: this.isLoading
      });
    });
  }
}

const authState = new AuthState();

// === SMS AUTHENTICATION SERVICE ===

export class SMSAuthService {
  constructor() {
    this.initializeAuthListener();
  }

  /**
   * Initialize Firebase auth state listener
   */
  initializeAuthListener() {
    onAuthStateChanged(auth, async (user) => {
      if (user) {
        try {
          // Get user profile from Firestore
          const userProfile = await this.getUserProfile(user.uid);
          
          const userData = {
            uid: user.uid,
            phoneNumber: userProfile?.phoneNumber || user.providerData[0]?.phoneNumber,
            country: userProfile?.country,
            role: userProfile?.role,
            isAdmin: userProfile?.isAdmin || false,
            displayName: userProfile?.displayName,
            email: userProfile?.email,
            createdAt: userProfile?.createdAt,
            lastLoginAt: new Date()
          };

          // Update last login time
          if (userProfile) {
            await this.updateUserProfile(user.uid, { lastLoginAt: Timestamp.now() });
          }

          authState.setUser(userData);
        } catch (error) {
          console.error('Error getting user profile:', error);
          authState.setUser({
            uid: user.uid,
            phoneNumber: user.providerData[0]?.phoneNumber,
            isAdmin: false
          });
        }
      } else {
        authState.setUser(null);
      }
    });
  }

  /**
   * Send OTP to phone number
   */
  async sendOTP(phoneNumber, country) {
    try {
      const sendOTPFunction = httpsCallable(functions, 'sendOTP');
      
      const result = await sendOTPFunction({
        phoneNumber: phoneNumber,
        country: country
      });

      return {
        success: true,
        message: result.data.message,
        expiresAt: result.data.expiresAt
      };
    } catch (error) {
      console.error('Send OTP Error:', error);
      
      // Handle different error types
      if (error.code === 'functions/resource-exhausted') {
        throw new Error('Please wait before requesting another OTP');
      } else if (error.code === 'functions/invalid-argument') {
        throw new Error('Invalid phone number format');
      } else {
        throw new Error(error.message || 'Failed to send OTP');
      }
    }
  }

  /**
   * Verify OTP and sign in user
   */
  async verifyOTP(phoneNumber, otp, country) {
    try {
      const verifyOTPFunction = httpsCallable(functions, 'verifyOTP');
      
      const result = await verifyOTPFunction({
        phoneNumber: phoneNumber,
        otp: otp,
        country: country
      });

      if (result.data.success) {
        // Sign in with custom token
        const userCredential = await signInWithCustomToken(auth, result.data.customToken);
        
        // Create or update user profile
        await this.createOrUpdateUserProfile(userCredential.user.uid, {
          phoneNumber: phoneNumber,
          country: country,
          lastLoginAt: Timestamp.now()
        });

        return {
          success: true,
          user: result.data.user,
          message: result.data.message
        };
      } else {
        throw new Error(result.data.message || 'OTP verification failed');
      }
    } catch (error) {
      console.error('Verify OTP Error:', error);
      
      if (error.code === 'functions/invalid-argument') {
        throw new Error('Invalid OTP or expired');
      } else {
        throw new Error(error.message || 'Failed to verify OTP');
      }
    }
  }

  /**
   * Admin login with email/password (fallback for admin users)
   */
  async adminLogin(email, password) {
    try {
      // For admin users, we can still use Firebase Auth email/password
      // This is more secure for admin access and the cost is minimal
      const { signInWithEmailAndPassword } = await import('firebase/auth');
      
      const userCredential = await signInWithEmailAndPassword(auth, email, password);
      
      // Verify admin status
      const adminProfile = await this.getAdminProfile(userCredential.user.uid);
      
      if (!adminProfile || !adminProfile.isAdmin) {
        await this.signOut();
        throw new Error('Access denied: Admin privileges required');
      }

      return {
        success: true,
        user: userCredential.user,
        adminData: adminProfile
      };
    } catch (error) {
      console.error('Admin login error:', error);
      throw new Error(error.message || 'Admin login failed');
    }
  }

  /**
   * Sign out user
   */
  async signOut() {
    try {
      await firebaseSignOut(auth);
      authState.setUser(null);
      return { success: true };
    } catch (error) {
      console.error('Sign out error:', error);
      throw new Error('Failed to sign out');
    }
  }

  /**
   * Get current user
   */
  getCurrentUser() {
    return authState.user;
  }

  /**
   * Check if user is authenticated
   */
  isAuthenticated() {
    return !!authState.user;
  }

  /**
   * Check if user is admin
   */
  isAdmin() {
    return authState.user?.isAdmin || false;
  }

  /**
   * Subscribe to auth state changes
   */
  onAuthStateChanged(callback) {
    return authState.addListener(callback);
  }

  /**
   * Get user profile from Firestore
   */
  async getUserProfile(uid) {
    try {
      const userDoc = await getDoc(doc(db, 'users', uid));
      return userDoc.exists() ? userDoc.data() : null;
    } catch (error) {
      console.error('Error getting user profile:', error);
      return null;
    }
  }

  /**
   * Get admin profile from Firestore
   */
  async getAdminProfile(uid) {
    try {
      const adminDoc = await getDoc(doc(db, 'admin_users', uid));
      return adminDoc.exists() ? adminDoc.data() : null;
    } catch (error) {
      console.error('Error getting admin profile:', error);
      return null;
    }
  }

  /**
   * Create or update user profile
   */
  async createOrUpdateUserProfile(uid, data) {
    try {
      const userRef = doc(db, 'users', uid);
      const userDoc = await getDoc(userRef);

      if (userDoc.exists()) {
        // Update existing profile
        await updateDoc(userRef, {
          ...data,
          updatedAt: Timestamp.now()
        });
      } else {
        // Create new profile
        await setDoc(userRef, {
          ...data,
          uid: uid,
          createdAt: Timestamp.now(),
          updatedAt: Timestamp.now(),
          isVerified: true // Phone number is verified through OTP
        });
      }
    } catch (error) {
      console.error('Error creating/updating user profile:', error);
      throw error;
    }
  }

  /**
   * Update user profile
   */
  async updateUserProfile(uid, data) {
    try {
      const userRef = doc(db, 'users', uid);
      await updateDoc(userRef, {
        ...data,
        updatedAt: Timestamp.now()
      });
    } catch (error) {
      console.error('Error updating user profile:', error);
      throw error;
    }
  }

  /**
   * Check if phone number is already registered
   */
  async isPhoneNumberRegistered(phoneNumber, country) {
    try {
      const usersQuery = query(
        collection(db, 'users'),
        where('phoneNumber', '==', phoneNumber),
        where('country', '==', country)
      );
      
      const snapshot = await getDocs(usersQuery);
      return !snapshot.empty;
    } catch (error) {
      console.error('Error checking phone number:', error);
      return false;
    }
  }

  /**
   * Register new user with phone number
   */
  async registerUser(phoneNumber, country, additionalData = {}) {
    try {
      // Check if phone number is already registered
      const isRegistered = await this.isPhoneNumberRegistered(phoneNumber, country);
      
      if (isRegistered) {
        throw new Error('Phone number is already registered');
      }

      // Send OTP for verification
      const otpResult = await this.sendOTP(phoneNumber, country);
      
      return {
        success: true,
        message: 'Please verify your phone number with the OTP sent',
        expiresAt: otpResult.expiresAt,
        nextStep: 'verify_otp'
      };
    } catch (error) {
      console.error('Register user error:', error);
      throw error;
    }
  }

  /**
   * Complete user registration after OTP verification
   */
  async completeRegistration(phoneNumber, otp, country, userData = {}) {
    try {
      // Verify OTP and create account
      const verifyResult = await this.verifyOTP(phoneNumber, otp, country);
      
      if (verifyResult.success) {
        // Update user profile with additional data
        if (Object.keys(userData).length > 0) {
          await this.updateUserProfile(auth.currentUser.uid, userData);
        }

        return {
          success: true,
          message: 'Registration completed successfully',
          user: verifyResult.user
        };
      }
      
      throw new Error('Registration failed');
    } catch (error) {
      console.error('Complete registration error:', error);
      throw error;
    }
  }

  /**
   * Test SMS configuration (Admin only)
   */
  async testSMSConfig(phoneNumber, message, country, provider, configuration) {
    try {
      if (!this.isAdmin()) {
        throw new Error('Admin access required');
      }

      const testSMSFunction = httpsCallable(functions, 'testSMSConfig');
      
      const result = await testSMSFunction({
        phoneNumber: phoneNumber,
        message: message,
        country: country,
        provider: provider,
        configuration: configuration
      });

      return result.data;
    } catch (error) {
      console.error('Test SMS config error:', error);
      throw new Error(error.message || 'Failed to test SMS configuration');
    }
  }

  /**
   * Get SMS statistics (Admin only)
   */
  async getSMSStatistics(country) {
    try {
      if (!this.isAdmin()) {
        throw new Error('Admin access required');
      }

      const getStatsFunction = httpsCallable(functions, 'getSMSStatistics');
      
      const result = await getStatsFunction({ country: country });
      
      return result.data;
    } catch (error) {
      console.error('Get SMS statistics error:', error);
      throw new Error(error.message || 'Failed to get SMS statistics');
    }
  }
}

// === EXPORT DEFAULT INSTANCE ===

const smsAuthService = new SMSAuthService();

export default smsAuthService;

// === UTILITY FUNCTIONS ===

/**
 * Format phone number for international use
 */
export const formatPhoneNumber = (phoneNumber, countryCode) => {
  // Remove all non-digits
  const digits = phoneNumber.replace(/\D/g, '');
  
  // Add country code if not present
  if (!digits.startsWith(countryCode)) {
    return `+${countryCode}${digits}`;
  }
  
  return `+${digits}`;
};

/**
 * Validate phone number format
 */
export const isValidPhoneNumber = (phoneNumber) => {
  const phoneRegex = /^\+[1-9]\d{1,14}$/;
  return phoneRegex.test(phoneNumber);
};

/**
 * Get country code from phone number
 */
export const getCountryCodeFromPhone = (phoneNumber) => {
  // Simple country code extraction (you might want to use a library like libphonenumber)
  const countryMappings = {
    '+1': 'US',
    '+44': 'GB', 
    '+91': 'IN',
    '+94': 'LK',
    '+61': 'AU',
    '+86': 'CN',
    '+49': 'DE',
    '+33': 'FR',
    '+81': 'JP'
  };
  
  for (const [code, country] of Object.entries(countryMappings)) {
    if (phoneNumber.startsWith(code)) {
      return { code: code.substring(1), country };
    }
  }
  
  return { code: null, country: null };
};

/**
 * Generate OTP (for testing purposes)
 */
export const generateTestOTP = () => {
  return Math.floor(100000 + Math.random() * 900000).toString();
};
